core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointService.java

```diff
@@ -32,33 +33,41 @@
  */
 public class LocalCheckpointService extends AbstractIndexShardComponent {
 
+    public static String MAX_SEQ_NO = "max_seq_no";
+    public static String LOCAL_CHECKPOINT_KEY = "local_checkpoint";
+
     /**
      * we keep a bit for each seq No that is still pending. to optimize allocation, we do so in multiple arrays
      * allocating them on demand and cleaning up while completed. This setting controls the size of the arrays
      */
     public static Setting<Integer> SETTINGS_BIT_ARRAYS_SIZE = Setting.intSetting("index.seq_no.checkpoint.bit_arrays_size", 1024,
         4, Setting.Property.IndexScope);
 
-
     /**
      * an ordered list of bit arrays representing pending seq nos. The list is "anchored" in {@link #firstProcessedSeqNo}
      * which marks the seqNo the fist bit in the first array corresponds to.
      */
     final LinkedList<FixedBitSet> processedSeqNo;
     final int bitArraysSize;
-    long firstProcessedSeqNo = 0;
+    long firstProcessedSeqNo;
 
     /** the current local checkpoint, i.e., all seqNo lower (&lt;=) than this number have been completed */
-    volatile long checkpoint = SequenceNumbersService.NO_OPS_PERFORMED;
+    volatile long checkpoint;
 
     /** the next available seqNo - used for seqNo generation */
-    volatile long nextSeqNo = 0;
+    volatile long nextSeqNo;
 
+    public LocalCheckpointService(final ShardId shardId, final IndexSettings indexSettings) {
+        this(shardId, indexSettings, SequenceNumbersService.NO_OPS_PERFORMED, SequenceNumbersService.NO_OPS_PERFORMED);
+    }
 
-    public LocalCheckpointService(ShardId shardId, IndexSettings indexSettings) {
+    public LocalCheckpointService(final ShardId shardId, final IndexSettings indexSettings, final long maxSeqNo, final long checkpoint) {
         super(shardId, indexSettings);
         bitArraysSize = SETTINGS_BIT_ARRAYS_SIZE.get(indexSettings.getSettings());
         processedSeqNo = new LinkedList<>();
+        firstProcessedSeqNo = checkpoint + 1;
```

Comment:
what happens if checkpoint is set to `NO_OPS_PERFORMED`? I think you get a 0 which is fine, but I rather make it explicit rather than rely on `NO_OPS_PERFORMED` being -1.